servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/TimeoutHttpRequesterFilter.java

```diff
@@ -174,10 +172,37 @@ public StreamingHttpConnectionFilter create(final FilterableStreamingHttpConnect
         return new StreamingHttpConnectionFilter(connection) {
             @Override
             public Single<StreamingHttpResponse> request(final StreamingHttpRequest request) {
-                HttpExecutionContext executionContext = connection.executionContext();
-                return TimeoutHttpRequesterFilter.this.withTimeout(request, r -> delegate().request(r),
-                        executionContext.executionStrategy().hasOffloads() ?
-                                executionContext.executor() : executionContext.ioExecutor());
+                final FilterableStreamingHttpConnection delegate = delegate();
+                return TimeoutHttpRequesterFilter.this.withTimeout(request,
+                        delegate.connectionContext().protocol().major() > 1 ? delegate::request :
+                                r -> delegate.request(r).liftSync(new BeforeFinallyHttpOperator(
+                                        new TerminalSignalConsumer() {
+                                            @Override
+                                            public void onComplete() {
+                                                // noop
+                                            }
+
+                                            @Override
+                                            public void onError(final Throwable throwable) {
+                                                // noop
+                                            }
+
+                                            @Override
+                                            public void cancel() {
+                                                // In alignment with cancellation processing in
+                                                // LoadBalancedStreamingHttpClient, we need to close the HTTP/1.x
+                                                // connection when this filter is applied at connection level.
+                                                // Otherwise, a connection will be marked as "free" and can be selected
+                                                // for another request, racing with closure.
+                                                delegate.closeAsync().subscribe();
```

Comment:
Aims of the biggest challenges is identifying who initiates connection closure and why. I know our api doesn’t support exception/rational propagation but should we add a debug log statement and an internal utility method to share code with the other spot that does this?

Reply:
Agreed, it's a bit unfortunate that `cancel` doesn't carry any context with it. 
There is no good place for a utility (and the line is really simple) bcz this filter and `LoadBalancedStreamingHttpClient` are in different modules. I will add debug logging.